File: ios/Classes/PeerManager.swift
import Foundation
import CoreBluetooth

/// Manages discovered and connected peers
class PeerManager {
    private let tag = "PeerManager"

    // Thread-safe dictionaries for peers
    private let queue = DispatchQueue(label: "com.ble_mesh.peermanager")
    private var discoveredPeers: [String: Peer] = [:]
    private var connectedPeers: [String: Peer] = [:]

    // Callbacks
    var onPeerDiscovered: ((Peer) -> Void)?
    var onPeerConnected: ((Peer) -> Void)?
    var onPeerDisconnected: ((Peer) -> Void)?

    /// Add a discovered peer
    func addDiscoveredPeer(_ peer: Peer) {
        queue.async { [weak self] in
            guard let self = self else { return }

            if let existingPeer = self.discoveredPeers[peer.id] {
                // Update existing peer (RSSI, lastSeen, etc.)
                let updatedPeer = Peer(
                    senderId: existingPeer.senderId,
                    connectionId: peer.id,
                    nickname: peer.nickname,
                    rssi: peer.rssi,
                    lastSeen: Date(),
                    connectionState: existingPeer.connectionState,
                    hopCount: existingPeer.hopCount,
                    lastForwardTime: existingPeer.lastForwardTime,
                    isBlocked: existingPeer.isBlocked,
                    peripheral: peer.peripheral
                )
                self.discoveredPeers[peer.id] = updatedPeer
                print("[\(self.tag)] Updated peer: \(peer.id) (\(peer.nickname)), RSSI: \(peer.rssi)")
            } else {
                // New peer discovered
                self.discoveredPeers[peer.id] = peer
                print("[\(self.tag)] New peer discovered: \(peer.id) (\(peer.nickname))")

                DispatchQueue.main.async {
                    self.onPeerDiscovered?(peer)
                }
            }
        }
    }

    /// Mark a peer as connected
    func markPeerConnected(_ peerId: String) {
        queue.async { [weak self] in
            guard let self = self else { return }
            guard let peer = self.discoveredPeers[peerId] else { return }

            let connectedPeer = Peer(
                senderId: peer.senderId,
                connectionId: peer.id,
                nickname: peer.nickname,
                rssi: peer.rssi,
                lastSeen: Date(),
                connectionState: .connected,
                hopCount: peer.hopCount,
                lastForwardTime: peer.lastForwardTime,
                isBlocked: peer.isBlocked,
                peripheral: peer.peripheral
            )

            self.connectedPeers[peerId] = connectedPeer
            self.discoveredPeers[peerId] = connectedPeer

            print("[\(self.tag)] Peer connected: \(peerId) (\(peer.nickname))")

            DispatchQueue.main.async {
                self.onPeerConnected?(connectedPeer)
            }
        }
    }

    /// Mark a peer as disconnected
    func markPeerDisconnected(_ peerId: String) {
        queue.async { [weak self] in
            guard let self = self else { return }
            guard let peer = self.connectedPeers.removeValue(forKey: peerId) else { return }

            let disconnectedPeer = Peer(
                senderId: peer.senderId,
                connectionId: peer.id,
                nickname: peer.nickname,
                rssi: peer.rssi,
                lastSeen: Date(),
                connectionState: .disconnected,
                hopCount: peer.hopCount,
                lastForwardTime: peer.lastForwardTime,
                isBlocked: peer.isBlocked,
                peripheral: peer.peripheral
            )

            self.discoveredPeers[peerId] = disconnectedPeer

            print("[\(self.tag)] Peer disconnected: \(peerId) (\(peer.nickname))")

            DispatchQueue.main.async {
                self.onPeerDisconnected?(disconnectedPeer)
            }
        }
    }

    /// Remove a peer completely
    func removePeer(_ peerId: String) {
        queue.async { [weak self] in
            guard let self = self else { return }
            self.discoveredPeers.removeValue(forKey: peerId)
            self.connectedPeers.removeValue(forKey: peerId)
            print("[\(self.tag)] Peer removed: \(peerId)")
        }
    }

    /// Get a peer by ID (connectionId)
    func getPeer(_ peerId: String) -> Peer? {
        return queue.sync {
            return discoveredPeers[peerId]
        }
    }

    /// Get a peer by senderId (stable identifier)
    func getPeerBySenderId(_ senderId: String) -> Peer? {
        return queue.sync {
            return discoveredPeers.values.first { $0.senderId == senderId }
        }
    }

    /// Get connectionId by senderId
    func getConnectionIdBySenderId(_ senderId: String) -> String? {
        return queue.sync {
            return discoveredPeers.values.first { $0.senderId == senderId }?.connectionId
        }
    }

    /// Update peer's senderId
    func updatePeerSenderId(_ connectionId: String, senderId: String) {
        queue.async { [weak self] in
            guard let self = self else { return }
            guard var peer = self.discoveredPeers[connectionId] else { return }

            // Create updated peer with new senderId
            let updatedPeer = Peer(
                senderId: senderId,
                connectionId: peer.connectionId,
                nickname: peer.nickname,
                rssi: peer.rssi,
                lastSeen: peer.lastSeen,
                connectionState: peer.connectionState,
                hopCount: peer.hopCount,
                lastForwardTime: peer.lastForwardTime,
                isBlocked: peer.isBlocked,
                peripheral: peer.peripheral
            )

            self.discoveredPeers[connectionId] = updatedPeer

            // Also update in connectedPeers if present
            if self.connectedPeers[connectionId] != nil {
                self.connectedPeers[connectionId] = updatedPeer
            }

            print("[\\(self.tag)] Updated senderId for peer \\(connectionId): \\(senderId)")
        }
    }

    /// Update peer's connection state
    func updatePeerConnectionState(_ senderId: String, state: PeerConnectionState) {
        queue.async { [weak self] in
            guard let self = self else { return }
            guard let peer = self.getPeerBySenderId(senderId) else { return }

            let updatedPeer = Peer(
                senderId: peer.senderId,
                connectionId: peer.connectionId,
                nickname: peer.nickname,
                rssi: peer.rssi,
                lastSeen: Date(),
                connectionState: state,
                hopCount: peer.hopCount,
                lastForwardTime: peer.lastForwardTime,
                isBlocked: peer.isBlocked,
                peripheral: peer.peripheral
            )

            self.discoveredPeers[peer.connectionId] = updatedPeer

            // Manage connectedPeers based on state
            if state == .connected {
                self.connectedPeers[peer.connectionId] = updatedPeer
            } else if state == .disconnected {
                self.connectedPeers.removeValue(forKey: peer.connectionId)
            }

            print("[\\(self.tag)] Updated connection state for senderId \\(senderId): \\(state.rawValue)")
        }
    }

    /// Get all discovered peers
    func getDiscoveredPeers() -> [Peer] {
        return queue.sync {
            return Array(discoveredPeers.values)
        }
    }

    /// Get all connected peers
    func getConnectedPeers() -> [Peer] {
        return queue.sync {
            return Array(connectedPeers.values)
        }
    }

    /// Get count of connected peers
    func getConnectedPeerCount() -> Int {
        return queue.sync {
            return connectedPeers.count
        }
    }

    /// Check if a peer is connected
    func isPeerConnected(_ peerId: String) -> Bool {
        return queue.sync {
            return connectedPeers[peerId] != nil
        }
    }

    /// Clear all peers
    func clearAll() {
        queue.async { [weak self] in
            guard let self = self else { return }
            self.discoveredPeers.removeAll()
            self.connectedPeers.removeAll()
            print("[\(self.tag)] All peers cleared")
        }
    }

    /// Remove stale peers (not seen for a while)
    func removeStalePeers(timeoutMs: TimeInterval = 60.0) {
        queue.async { [weak self] in
            guard let self = self else { return }

            let currentTime = Date()
            var staleCount = 0

            let stalePeers = self.discoveredPeers.filter { (_, peer) in
                return peer.connectionState != .connected && currentTime.timeIntervalSince(peer.lastSeen) > timeoutMs
            }

            for (peerId, _) in stalePeers {
                self.discoveredPeers.removeValue(forKey: peerId)
                staleCount += 1
            }

            if staleCount > 0 {
                print("[\(self.tag)] Removed \(staleCount) stale peers")
            }
        }
    }

    /// Clean up resources
    func cleanup() {
        clearAll()
        onPeerDiscovered = nil
        onPeerConnected = nil
        onPeerDisconnected = nil
    }
}

File: android/src/main/kotlin/com/ble_mesh/PeerManager.kt
package com.ble_mesh

import android.util.Log
import com.ble_mesh.models.Peer
import com.ble_mesh.models.PeerConnectionState
import java.util.concurrent.ConcurrentHashMap

/**
 * Manages discovered and connected peers
 */
class PeerManager {
    private val tag = "PeerManager"

    // Map of peer ID to Peer
    private val discoveredPeers = ConcurrentHashMap<String, Peer>()
    private val connectedPeers = ConcurrentHashMap<String, Peer>()

    // Callbacks
    var onPeerDiscovered: ((Peer) -> Unit)? = null
    var onPeerConnected: ((Peer) -> Unit)? = null
    var onPeerDisconnected: ((Peer) -> Unit)? = null

    /**
     * Add a discovered peer
     */
    fun addDiscoveredPeer(peer: Peer) {
        val existingPeer = discoveredPeers[peer.id]

        if (existingPeer == null) {
            // New peer discovered
            discoveredPeers[peer.id] = peer
            Log.d(tag, "New peer discovered: ${peer.id} (${peer.nickname})")
            onPeerDiscovered?.invoke(peer)
        } else {
            // Update existing peer (RSSI, lastSeen, etc.)
            val updatedPeer = peer.copy(lastSeen = System.currentTimeMillis())
            discoveredPeers[peer.id] = updatedPeer
            Log.d(tag, "Updated peer: ${peer.id} (${peer.nickname}), RSSI: ${peer.rssi}")
        }
    }

    /**
     * Mark a peer as connected
     */
    fun markPeerConnected(peerId: String) {
        val peer = discoveredPeers[peerId] ?: return

        val connectedPeer = peer.copy(
            connectionState = PeerConnectionState.CONNECTED,
            lastSeen = System.currentTimeMillis()
        )

        connectedPeers[peerId] = connectedPeer
        discoveredPeers[peerId] = connectedPeer

        Log.d(tag, "Peer connected: $peerId (${peer.nickname})")
        onPeerConnected?.invoke(connectedPeer)
    }

    /**
     * Mark a peer as disconnected
     */
    fun markPeerDisconnected(peerId: String) {
        val peer = connectedPeers.remove(peerId) ?: return

        val disconnectedPeer = peer.copy(
            connectionState = PeerConnectionState.DISCONNECTED,
            lastSeen = System.currentTimeMillis()
        )

        discoveredPeers[peerId] = disconnectedPeer

        Log.d(tag, "Peer disconnected: $peerId (${peer.nickname})")
        onPeerDisconnected?.invoke(disconnectedPeer)
    }

    /**
     * Remove a peer completely
     */
    fun removePeer(peerId: String) {
        discoveredPeers.remove(peerId)
        connectedPeers.remove(peerId)
        Log.d(tag, "Peer removed: $peerId")
    }

    /**
     * Get a peer by ID (connectionId)
     */
    fun getPeer(peerId: String): Peer? {
        return discoveredPeers[peerId]
    }

    /**
     * Get a peer by sender UUID
     */
    fun getPeerBySenderId(senderId: String): Peer? {
        return discoveredPeers.values.firstOrNull { it.senderId == senderId }
    }

    /**
     * Update peer senderId (after handshake)
     */
    fun updatePeerSenderId(connectionId: String, senderId: String) {
        discoveredPeers[connectionId]?.let { peer ->
            val updatedPeer = peer.copy(senderId = senderId)
            discoveredPeers[connectionId] = updatedPeer
            if (connectedPeers.containsKey(connectionId)) {
                connectedPeers[connectionId] = updatedPeer
            }
            Log.d(tag, "Updated peer senderId: $connectionId -> $senderId")
        }
    }

    /**
     * Update peer connection state
     */
    fun updatePeerConnectionState(senderId: String, state: com.ble_mesh.models.PeerConnectionState) {
        val peer = getPeerBySenderId(senderId) ?: return
        val updatedPeer = peer.copy(
            connectionState = state,
            lastSeen = System.currentTimeMillis()
        )
        discoveredPeers[peer.connectionId] = updatedPeer
        if (state == com.ble_mesh.models.PeerConnectionState.CONNECTED) {
            connectedPeers[peer.connectionId] = updatedPeer
        } else if (state == com.ble_mesh.models.PeerConnectionState.DISCONNECTED) {
            connectedPeers.remove(peer.connectionId)
        }
        Log.d(tag, "Updated peer state: $senderId -> ${state.name}")
    }

    /**
     * Get all discovered peers
     */
    fun getDiscoveredPeers(): List<Peer> {
        return discoveredPeers.values.toList()
    }

    /**
     * Get all connected peers
     */
    fun getConnectedPeers(): List<Peer> {
        return connectedPeers.values.toList()
    }

    /**
     * Get count of connected peers
     */
    fun getConnectedPeerCount(): Int {
        return connectedPeers.size
    }

    /**
     * Check if a peer is connected
     */
    fun isPeerConnected(peerId: String): Boolean {
        return connectedPeers.containsKey(peerId)
    }

    /**
     * Clear all peers
     */
    fun clearAll() {
        discoveredPeers.clear()
        connectedPeers.clear()
        Log.d(tag, "All peers cleared")
    }

    /**
     * Remove stale peers (not seen for a while)
     */
    fun removeStalePeers(timeoutMs: Long = 60000L) {
        val currentTime = System.currentTimeMillis()
        val stalePeers = discoveredPeers.filter { (_, peer) ->
            peer.connectionState != PeerConnectionState.CONNECTED && (currentTime - peer.lastSeen) > timeoutMs
        }

        stalePeers.forEach { (peerId, _) ->
            removePeer(peerId)
        }

        if (stalePeers.isNotEmpty()) {
            Log.d(tag, "Removed ${stalePeers.size} stale peers")
        }
    }

    /**
     * Clean up resources
     */
    fun cleanup() {
        clearAll()
        onPeerDiscovered = null
        onPeerConnected = null
        onPeerDisconnected = null
    }
}

